File: algorithm/study/Core-0D-MST-Topo/1197/cpp/src/solution.cpp
#include <bits/stdc++.h>
using namespace std;
int par[10001], rnk[10001];
int find(int x){ return par[x]==x? x : par[x]=find(par[x]); }
bool unite(int a,int b){
    a=find(a); b=find(b);
    if(a==b) return false;
    if(rnk[a]<rnk[b]) swap(a,b);
    par[b]=a;
    if(rnk[a]==rnk[b]) rnk[a]++;
    return true;
}
int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int V,E; cin >> V >> E;
    vector<tuple<int,int,int>> edges(E);
    for(auto &[w,u,v]:edges) cin >> u >> v >> w;
    sort(edges.begin(), edges.end());
    for(int i=1;i<=V;i++) par[i]=i;
    long long total=0; int cnt=0;
    for(auto &[w,u,v]:edges){
        if(unite(u,v)){
            total+=w;
            if(++cnt==V-1) break;
        }
    }
    cout << total << '\n';
    return 0;
}

File: algorithm/study/Core-0D-MST-Topo/1197/python/src/solution.py
import sys
input = sys.stdin.readline

def find(parent, x):
    while parent[x] != x:
        parent[x] = parent[parent[x]]
        x = parent[x]
    return x

def union(parent, rank, a, b):
    a, b = find(parent, a), find(parent, b)
    if a == b:
        return False
    if rank[a] < rank[b]:
        a, b = b, a
    parent[b] = a
    if rank[a] == rank[b]:
        rank[a] += 1
    return True

def main():
    v, e = map(int, input().split())
    edges = []
    for _ in range(e):
        a, b, c = map(int, input().split())
        edges.append((c, a, b))
    edges.sort()
    parent = list(range(v + 1))
    rank = [0] * (v + 1)
    total = 0
    cnt = 0
    for w, a, b in edges:
        if union(parent, rank, a, b):
            total += w
            cnt += 1
            if cnt == v - 1:
                break
    print(total)

if __name__ == "__main__":
    main()
